server/src/main/java/org/opensearch/wlm/cancellation/QueryGroupTaskCancellationService.java

```diff
@@ -47,46 +53,80 @@
  */
 public class QueryGroupTaskCancellationService {
     public static final double MIN_VALUE = 1e-9;
+    private static final Logger log = LogManager.getLogger(QueryGroupTaskCancellationService.class);
 
     private final WorkloadManagementSettings workloadManagementSettings;
     private final TaskSelectionStrategy taskSelectionStrategy;
     private final QueryGroupResourceUsageTrackerService resourceUsageTrackerService;
     // a map of QueryGroupId to its corresponding QueryGroupLevelResourceUsageView object
     Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews;
-    private final Collection<QueryGroup> activeQueryGroups;
-    private final Collection<QueryGroup> deletedQueryGroups;
+    private Function<String, QueryGroupState> queryGroupStateAccessor;
 
     public QueryGroupTaskCancellationService(
         WorkloadManagementSettings workloadManagementSettings,
         TaskSelectionStrategy taskSelectionStrategy,
-        QueryGroupResourceUsageTrackerService resourceUsageTrackerService,
-        Collection<QueryGroup> activeQueryGroups,
-        Collection<QueryGroup> deletedQueryGroups
+        QueryGroupResourceUsageTrackerService resourceUsageTrackerService
     ) {
         this.workloadManagementSettings = workloadManagementSettings;
         this.taskSelectionStrategy = taskSelectionStrategy;
         this.resourceUsageTrackerService = resourceUsageTrackerService;
-        this.activeQueryGroups = activeQueryGroups;
-        this.deletedQueryGroups = deletedQueryGroups;
+    }
+
+    public void setQueryGroupStateMapAccessor(final Function<String, QueryGroupState> queryGroupStateAccessor) {
+        this.queryGroupStateAccessor = queryGroupStateAccessor;
     }
```

Comment:
This function will not change over the lifecycle of `QueryGroupTaskCancellationService`, right? Should not have setter method in that case, should be provided using constructor

Reply:
I think that's fair, let me inject this into the constructor itself